behaviour/graql/language/define.feature

```diff
@@ -1018,92 +928,700 @@ Feature: Graql Define Query
+      };
+      """
 
 
-  #########
-  # RULES #
-  #########
+  Scenario: define a rule with no `then` clause throws
+    Then graql define throws
+      """
+      define
+      nickname sub name;
+      person has nickname;
+      robert sub rule,
+      when {
+        $p has name "Robert";
+      };
+      """
 
-  Scenario: define a rule creates a rule
 
-  Scenario: define a rule with nested negation throws on commit
+  Scenario: define a rule with an empty `when` clause throws
+    Then graql define throws
+      """
+      define
+      nickname sub name;
+      person has nickname;
+      has-nickname-bob sub rule,
+      when {
+      }, then {
+        $p has nickname "Bob";
+      };
+      """
 
-  Scenario: define a rule with two conclusions throws on commit
 
-  Scenario: define a rule with disjunction throws on commit
+  Scenario: define a rule with an empty `then` clause throws
+    Then graql define throws
+      """
+      define
+      nickname sub name;
+      person has nickname;
+      robert sub rule,
+      when {
+        $p has name "Robert";
+      }, then {
+      };
+      """
 
-  Scenario: define rule with an unbound variable in the `then` throws on commit
 
-  Scenario: define a non-insertable `then` throws on commit (eg. missing specific roles, or attribute value)
+  Scenario: define a rule with negation creates a rule
+    Given graql define
+      """
+      define
+      only-child sub attribute, value boolean;
+      siblings sub relation, relates sibling;
+      person plays sibling, has only-child;
+      only-child-rule sub rule,
+      when {
+        $p isa person;
+        not {
+          (sibling: $p, sibling: $p2) isa siblings;
+        };
+      }, then {
+        $p has only-child true;
+      };
+      """
+    Given the integrity is validated
+    When get answers of graql query
+      """
+      match $x sub rule; get;
+      """
+    Then concept identifiers are
+      |     | check | value           |
+      | ONL | label | only-child-rule |
+      | RUL | label | rule            |
+    Then uniquely identify answer concepts
+      | x   |
+      | ONL |
+      | RUL |
+
+
+  # TODO: better scenario description? what exactly does 'unbound negated pattern variable' mean?
+  Scenario: define a rule with unbound negated pattern variables throws
+    Then graql define throws
+      """
+      define
+      has-robert sub attribute, value boolean;
+      register sub entity, has has-robert;
+      register-has-no-robert sub rule,
+      when {
+        $register isa register;
+        not {
+          $p isa person, has name "Robert";
+        };
+      }, then {
+        $register has has-robert false;
+      };
+      """
+
 
-  Scenario: define a rule causing a loop throws on commit (eg. conclusion is negated in the `when`)
+  @ignore
+  # TODO: re-enable when all rules with nested negations throw on commit
+  Scenario: define a rule with nested negation throws on commit
```

Comment:
this isn't expected to throw - not(not()) is fine (we normalise all the queries to disjunctive normal form, so we flatten the double negated form), not($x has...; not ($x... $a)) should throw though

Reply:
Okay, here's a new case that should throw.
```
define
      nickname sub name;
      person has nickname;
      unemployed-robert-maybe-doesnt-not-have-nickname-bob sub rule,
      when {
        $p isa person;
        not {
          (employee: $p) isa employment;
          not {
            $p has name "Robert";
          };
        };
      }, then {
        $p has nickname "Bob";
      };```

Reply:
Just documenting the results of the call

- Unwanted error most likely triggered because `nickname sub name`, could be fixed by making `nickname sub attribute, value string`

Reply:
Updated to `nickname sub attribute, value string`.

The new error is:
```
INVALID_ARGUMENT: InvalidKBException-A structural validation error has occurred. Please correct the [`1`] errors found. 
The rule [unemployed-robert-maybe-doesnt-not-have-nickname-bob] contains a nested negation block
```
which is the correct behaviour!